common/context/src/main/java/io/helidon/common/context/ContextSingleton.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2024 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.context;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/**
+ * This is similar to {@link java.lang.ThreadLocal}, except it uses {@code static} context to store a value.
+ * Static context is either a context available under special classifier {@value #STATIC_CONTEXT_CLASSIFIER}, or the
+ * {@link io.helidon.common.context.Contexts#globalContext()} (if the qualifier is not registered).
+ * <p>
+ * The intention of this type is to have singletons for production runtime, while maintaining a separation during
+ * testing, where Helidon test extensions run each part of the test class in a test context.
+ * <p>
+ * Note that the instance is not simply registered in the context, but it uses this class, and the "owner type" as a classifier
```

Comment:
Here we describe a contract for the uniqueness which contradicts the semantic of a singleton.
I think we should remove that sentence and leave this as an implementation detail.